File: musiclibrary/src/main/java/com/lzx/musiclibrary/playback/player/ExoPlayback.java
package com.lzx.musiclibrary.playback.player;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioManager;
import android.net.Uri;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.danikula.videocache.HttpProxyCacheServer;
import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.DefaultLoadControl;
import com.google.android.exoplayer2.DefaultRenderersFactory;
import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.ExoPlayerFactory;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.Timeline;
import com.google.android.exoplayer2.audio.AudioAttributes;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.MediaSourceEventListener;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.source.dash.DashMediaSource;
import com.google.android.exoplayer2.source.dash.DefaultDashChunkSource;
import com.google.android.exoplayer2.source.hls.HlsMediaSource;
import com.google.android.exoplayer2.source.smoothstreaming.DefaultSsChunkSource;
import com.google.android.exoplayer2.source.smoothstreaming.SsMediaSource;
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultBandwidthMeter;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory;
import com.google.android.exoplayer2.upstream.HttpDataSource;
import com.google.android.exoplayer2.upstream.TransferListener;
import com.google.android.exoplayer2.util.Util;
import com.lzx.musiclibrary.MusicService;
import com.lzx.musiclibrary.aidl.model.SongInfo;
import com.lzx.musiclibrary.cache.CacheConfig;
import com.lzx.musiclibrary.cache.CacheUtils;
import com.lzx.musiclibrary.constans.State;
import com.lzx.musiclibrary.manager.FocusAndLockManager;
import com.lzx.musiclibrary.utils.BaseUtil;

import static com.google.android.exoplayer2.C.CONTENT_TYPE_MUSIC;
import static com.google.android.exoplayer2.C.USAGE_MEDIA;
import static com.lzx.musiclibrary.manager.FocusAndLockManager.AUDIO_NO_FOCUS_CAN_DUCK;
import static com.lzx.musiclibrary.manager.FocusAndLockManager.AUDIO_NO_FOCUS_NO_DUCK;
import static com.lzx.musiclibrary.manager.FocusAndLockManager.VOLUME_DUCK;
import static com.lzx.musiclibrary.manager.FocusAndLockManager.VOLUME_NORMAL;

/**
 * Created by xian on 2018/1/20.
 */

public class ExoPlayback implements Playback, FocusAndLockManager.AudioFocusChangeListener {

    private boolean isOpenCacheWhenPlaying = false;
    private boolean mPlayOnFocusGain;
    private boolean mAudioNoisyReceiverRegistered;
    private String mCurrentMediaId; //当前播放的媒体id
    private SongInfo mCurrentMediaSongInfo;

    private SimpleExoPlayer mExoPlayer;
    private final ExoPlayerEventListener mEventListener = new ExoPlayerEventListener();
    private boolean mExoPlayerNullIsStopped = false;
    private boolean isGiveUpAudioFocusManager = false;
    private FocusAndLockManager mFocusAndLockManager;

    private Callback mCallback;
    private Context mContext;
    private DataSource.Factory mediaDataSourceFactory;
    protected String userAgent;
    private static final DefaultBandwidthMeter BANDWIDTH_METER = new DefaultBandwidthMeter();

    private HttpProxyCacheServer mProxyCacheServer;
    private HttpProxyCacheServer.Builder builder;

    public ExoPlayback(Context context, CacheConfig cacheConfig, boolean isGiveUpAudioFocusManager) {
        Context applicationContext = context.getApplicationContext();
        this.mContext = applicationContext;
        this.isGiveUpAudioFocusManager = isGiveUpAudioFocusManager;
        mFocusAndLockManager = new FocusAndLockManager(applicationContext, this);
        userAgent = Util.getUserAgent(mContext, "ExoPlayer");
        mediaDataSourceFactory = buildDataSourceFactory(true);

        builder = CacheUtils.createHttpProxyCacheServerBuilder(mContext, cacheConfig);
        if (cacheConfig != null && cacheConfig.isOpenCacheWhenPlaying()) {
            isOpenCacheWhenPlaying = true;
        }
        mProxyCacheServer = builder.build();
    }

    private final IntentFilter mAudioNoisyIntentFilter = new IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY);

    private final BroadcastReceiver mAudioNoisyReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (AudioManager.ACTION_AUDIO_BECOMING_NOISY.equals(intent.getAction())) {
                if (isPlaying()) {
                    Intent i = new Intent(context, MusicService.class);
                    mContext.startService(i);
                }
            }
        }
    };

    /**
     * 释放服务使用的资源进行播放，这主要只是WiFi锁本地播放。 如果请求，ExoPlayer实例也被释放。
     *
     * @param releasePlayer 指示播放器是否也应该被释放
     */
    private void releaseResources(boolean releasePlayer) {
        if (releasePlayer && mExoPlayer != null) {
            mExoPlayer.release();
            mExoPlayer.removeListener(mEventListener);
            mExoPlayer = null;
            mExoPlayerNullIsStopped = true;
            mPlayOnFocusGain = false;
        }
        mFocusAndLockManager.releaseWifiLock();
    }

    @Override
    public void start() {

    }

    @Override
    public void stop(boolean notifyListeners) {
        mFocusAndLockManager.giveUpAudioFocus();
        unregisterAudioNoisyReceiver();
        releaseResources(true);

    }

    @Override
    public void setState(int state) {

    }

    @Override
    public int getState() {
        //STATE_IDLE      没有任何媒体播放。
        //STATE_BUFFERING 无法立即从当前位置进行播放
        //STATE_READY     可以从当前位置立即进行播放。 如果  {@link #getPlayWhenReady（）}为true，立即播放，否则暂停。
        //STATE_ENDED     已经完成播放媒体。
        int state = State.STATE_NONE;
        if (mExoPlayer == null) {
            state = mExoPlayerNullIsStopped ? State.STATE_STOP : State.STATE_NONE;
        } else {
            switch (mExoPlayer.getPlaybackState()) {
                case Player.STATE_IDLE:
                    state = State.STATE_IDLE;
                    break;
                case Player.STATE_BUFFERING:
                    state = State.STATE_ASYNC_LOADING;
                    break;
                case Player.STATE_READY:
                    state = mExoPlayer.getPlayWhenReady() ? State.STATE_PLAYING : State.STATE_PAUSED;
                    break;
                case Player.STATE_ENDED:
                    state = State.STATE_ENDED;
                    break;
            }
        }
        return state;
    }

    @Override
    public boolean isConnected() {
        return true;
    }

    @Override
    public boolean isPlaying() {
        return mPlayOnFocusGain || (mExoPlayer != null && mExoPlayer.getPlayWhenReady());
    }

    @Override
    public long getCurrentStreamPosition() {
        return mExoPlayer != null ? mExoPlayer.getCurrentPosition() : 0;
    }

    @Override
    public long getBufferedPosition() {
        long bufferedPosition = mExoPlayer != null ? mExoPlayer.getBufferedPosition() : 0;
        long duration = mExoPlayer != null ? mExoPlayer.getDuration() : 0;
        bufferedPosition = bufferedPosition * 2;
        if (bufferedPosition > duration) {
            bufferedPosition = duration;
        }
        if (isOpenCacheWhenPlaying && mCurrentMediaSongInfo != null) {
            boolean fullyCached = mProxyCacheServer.isCached(mCurrentMediaSongInfo.getSongUrl());
            return fullyCached ? duration : bufferedPosition;
        } else {
            return bufferedPosition;
        }
    }

    @Override
    public void updateLastKnownStreamPosition() {

    }

    @Override
    public void play(SongInfo info) {
        mPlayOnFocusGain = true;
        mFocusAndLockManager.tryToGetAudioFocus();
        registerAudioNoisyReceiver();
        String mediaId = info.getSongId();
        boolean mediaHasChanged = !TextUtils.equals(mediaId, mCurrentMediaId);
        if (mediaHasChanged) {
            mCurrentMediaId = mediaId;
            mCurrentMediaSongInfo = info;
        }

        if (mediaHasChanged || mExoPlayer == null) {
            releaseResources(false); // release everything except the player

            String source = info.getSongUrl();
            if (source != null && BaseUtil.isOnLineSource(source)) {
                source = source.replaceAll(" ", "%20"); // Escape spaces for URLs
            }
            if (TextUtils.isEmpty(source)) {
                if (mCallback != null) {
                    mCallback.onError("song url is null");
                }
                return;
            }

            Uri playUri;
            if (BaseUtil.isOnLineSource(source)) {
                String proxyUrl;
                if (isOpenCacheWhenPlaying && (getMediaType(null, Uri.parse(source)) == C.TYPE_OTHER)) {
                    proxyUrl = mProxyCacheServer.getProxyUrl(source);

                } else {
                    proxyUrl = source;
                }
                playUri = Uri.parse(proxyUrl);
            } else {
                playUri = BaseUtil.getLocalSourceUri(source);
            }
            if (playUri == null) {
                if (mCallback != null) {
                    mCallback.onError("song uri is null");
                }
                return;
            }

            //LogUtil.i("isOpenCacheWhenPlaying = " + isOpenCacheWhenPlaying + " playUri = " + playUri.toString());

            if (mExoPlayer == null) {
                mExoPlayer = ExoPlayerFactory.newSimpleInstance(new DefaultRenderersFactory(mContext),
                        new DefaultTrackSelector(), new DefaultLoadControl());
                mExoPlayer.addListener(mEventListener);
            }

            final AudioAttributes audioAttributes = new AudioAttributes.Builder()
                    .setContentType(CONTENT_TYPE_MUSIC)
                    .setUsage(USAGE_MEDIA)
                    .build();
            mExoPlayer.setAudioAttributes(audioAttributes);

            MediaSource mediaSource = buildMediaSource(playUri, null, null, null);
            mExoPlayer.prepare(mediaSource);
            mFocusAndLockManager.acquireWifiLock();
        }
        configurePlayerState();
    }

    /**
     * 构建不同的MediaSource
     */
    private MediaSource buildMediaSource(
            Uri uri,
            String overrideExtension,
            @Nullable Handler handler,
            @Nullable MediaSourceEventListener listener) {
        @C.ContentType int type = getMediaType(overrideExtension, uri);
        switch (type) {
            case C.TYPE_DASH:
                return new DashMediaSource.Factory(
                        new DefaultDashChunkSource.Factory(mediaDataSourceFactory),
                        buildDataSourceFactory(false))
                        .createMediaSource(uri, handler, listener);
            case C.TYPE_SS:
                return new SsMediaSource.Factory(
                        new DefaultSsChunkSource.Factory(mediaDataSourceFactory),
                        buildDataSourceFactory(false))
                        .createMediaSource(uri, handler, listener);
            case C.TYPE_HLS:
                return new HlsMediaSource.Factory(mediaDataSourceFactory)
                        .createMediaSource(uri, handler, listener);
            case C.TYPE_OTHER:
                return new ExtractorMediaSource.Factory(mediaDataSourceFactory)
                        .createMediaSource(uri, handler, listener);
            default: {
                throw new IllegalStateException("Unsupported type: " + type);
            }
        }
    }

    /**
     * 获取播放类型
     */
    private int getMediaType(String overrideExtension, Uri uri) {
        @C.ContentType int type = TextUtils.isEmpty(overrideExtension) ? Util.inferContentType(uri)
                : Util.inferContentType("." + overrideExtension);
        return type;
    }

    private DataSource.Factory buildDataSourceFactory(boolean useBandwidthMeter) {
        return buildDataSourceFactory(useBandwidthMeter ? BANDWIDTH_METER : null);
    }

    private DataSource.Factory buildDataSourceFactory(TransferListener<? super DataSource> listener) {
        return new DefaultDataSourceFactory(mContext, listener, buildHttpDataSourceFactory(listener));
    }

    private HttpDataSource.Factory buildHttpDataSourceFactory(
            TransferListener<? super DataSource> listener) {
        return new DefaultHttpDataSourceFactory(userAgent, listener);
    }


    @Override
    public void pause() {
        if (mExoPlayer != null) {
            mExoPlayer.setPlayWhenReady(false);
        }
        releaseResources(false);
        unregisterAudioNoisyReceiver();
    }

    @Override
    public void seekTo(long position) {
        if (mExoPlayer != null) {
            registerAudioNoisyReceiver();
            mExoPlayer.seekTo(position);

        }
    }

    @Override
    public void setCurrentMediaId(String mediaId) {
        this.mCurrentMediaId = mediaId;
    }

    @Override
    public String getCurrentMediaId() {
        return mCurrentMediaId;
    }

    @Override
    public int getDuration() {
        return mExoPlayer != null ? (int) mExoPlayer.getDuration() : 0;
    }

    @Override
    public SongInfo getCurrentMediaSongInfo() {
        return mCurrentMediaSongInfo;
    }

    @Override
    public void openCacheWhenPlaying(boolean isOpen) {
        isOpenCacheWhenPlaying = isOpen;
    }

    @Override
    public void setPlaybackParameters(float speed, float pitch) {
        if (mExoPlayer != null) {
            mExoPlayer.setPlaybackParameters(new PlaybackParameters(speed, pitch));
        }
    }

    @Override
    public void setVolume(float audioVolume) {
        if (mExoPlayer != null) {
            mExoPlayer.setVolume(audioVolume);
        }
    }

    @Override
    public int getAudioSessionId() {
        if (mExoPlayer!=null){
            return mExoPlayer.getAudioSessionId();
        }
        return 0;
    }

    @Override
    public void setCallback(Callback callback) {
        this.mCallback = callback;
    }

    /**
     * 根据音频焦点设置重新配置播放器并启动/重新启动。 这种方法
     * 启动/重新启动ExoPlayer的实例尊重当前的音频焦点状态。 所以，如果我们
     * 有焦点，会正常播放; 如果我们没有重点，它会离开玩家
     * 暂停或将其设置为低音量，具体取决于当前焦点允许的内容
     * 设置。    
     */
    private void configurePlayerState() {
        if (mFocusAndLockManager.getCurrentAudioFocusState() == AUDIO_NO_FOCUS_NO_DUCK) {
            if (!isGiveUpAudioFocusManager) {
                pause();
            }
        } else {
            registerAudioNoisyReceiver();
            if (mFocusAndLockManager.getCurrentAudioFocusState() == AUDIO_NO_FOCUS_CAN_DUCK) {
                mExoPlayer.setVolume(VOLUME_DUCK);
            } else {
                mExoPlayer.setVolume(VOLUME_NORMAL);
            }
            if (mPlayOnFocusGain) {
                mExoPlayer.setPlayWhenReady(true);
                mPlayOnFocusGain = false;
            }
            if (mExoPlayerNullIsStopped) {
                mExoPlayerNullIsStopped = false;
            }
        }
    }

    private void registerAudioNoisyReceiver() {
        if (!mAudioNoisyReceiverRegistered) {
            mContext.registerReceiver(mAudioNoisyReceiver, mAudioNoisyIntentFilter);
            mAudioNoisyReceiverRegistered = true;
        }
    }

    private void unregisterAudioNoisyReceiver() {
        if (mAudioNoisyReceiverRegistered) {
            mContext.unregisterReceiver(mAudioNoisyReceiver);
            mAudioNoisyReceiverRegistered = false;
        }
    }

    @Override
    public void onAudioFocusLossTransient() {
        mPlayOnFocusGain = mExoPlayer != null && mExoPlayer.getPlayWhenReady();
    }

    @Override
    public void onAudioFocusChange() {
        if (mExoPlayer != null) {
            configurePlayerState();
        }
    }


    /**
     * ExoPlayer事件监听器
     */
    private final class ExoPlayerEventListener implements Player.EventListener {


        @Override
        public void onTimelineChanged(Timeline timeline, Object manifest, int reason) {

        }

        @Override
        public void onTracksChanged(TrackGroupArray trackGroups, TrackSelectionArray trackSelections) {

        }

        @Override
        public void onLoadingChanged(boolean isLoading) {

        }

        @Override
        public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
            if (mCallback != null) {
                switch (playbackState) {
                    case Player.STATE_IDLE:
                        mCallback.onPlaybackStatusChanged(State.STATE_IDLE);
                        break;
                    case Player.STATE_BUFFERING:
                        mCallback.onPlaybackStatusChanged(State.STATE_ASYNC_LOADING);
                        break;
                    case Player.STATE_READY:
                        mCallback.onPlaybackStatusChanged(playWhenReady ? State.STATE_PLAYING : State.STATE_PAUSED);
                        break;
                    case Player.STATE_ENDED:
                        mCallback.onPlayCompletion();
                        // mCallback.onPlaybackStatusChanged(State.STATE_ENDED);
                        break;
                }
            }
        }

        @Override
        public void onRepeatModeChanged(int repeatMode) {

        }

        @Override
        public void onShuffleModeEnabledChanged(boolean shuffleModeEnabled) {

        }

        @Override
        public void onPlayerError(ExoPlaybackException error) {
            final String what;
            switch (error.type) {
                case ExoPlaybackException.TYPE_SOURCE:
                    what = error.getSourceException().getMessage();
                    break;
                case ExoPlaybackException.TYPE_RENDERER:
                    what = error.getRendererException().getMessage();
                    break;
                case ExoPlaybackException.TYPE_UNEXPECTED:
                    what = error.getUnexpectedException().getMessage();
                    break;
                default:
                    what = "Unknown: " + error;
            }
            if (mCallback != null) {
                mCallback.onError("ExoPlayer error " + what);
            }
        }

        @Override
        public void onPositionDiscontinuity(int reason) {

        }


        @Override
        public void onPlaybackParametersChanged(PlaybackParameters playbackParameters) {

        }

        @Override
        public void onSeekProcessed() {

        }
    }


}

File: starrysky/src/main/java/com/lzx/starrysky/playback/ExoPlayback.kt
package com.lzx.starrysky.playback

import android.content.Context
import android.net.Uri
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.DefaultRenderersFactory.ExtensionRendererMode
import com.google.android.exoplayer2.audio.AudioAttributes
import com.google.android.exoplayer2.ext.rtmp.RtmpDataSourceFactory
import com.google.android.exoplayer2.extractor.DefaultExtractorsFactory
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.source.dash.DashMediaSource
import com.google.android.exoplayer2.source.hls.HlsMediaSource
import com.google.android.exoplayer2.source.rtsp.RtspMediaSource
import com.google.android.exoplayer2.source.smoothstreaming.SsMediaSource
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector.ParametersBuilder
import com.google.android.exoplayer2.upstream.DataSource
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.google.android.exoplayer2.upstream.DefaultHttpDataSource
import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory
import com.google.android.exoplayer2.upstream.cache.Cache
import com.google.android.exoplayer2.upstream.cache.CacheDataSource
import com.google.android.exoplayer2.util.Util
import com.lzx.starrysky.SongInfo
import com.lzx.starrysky.StarrySky
import com.lzx.starrysky.cache.ExoCache
import com.lzx.starrysky.cache.ICache
import com.lzx.starrysky.playback.Playback.Companion.STATE_BUFFERING
import com.lzx.starrysky.playback.Playback.Companion.STATE_ERROR
import com.lzx.starrysky.playback.Playback.Companion.STATE_IDLE
import com.lzx.starrysky.playback.Playback.Companion.STATE_PAUSED
import com.lzx.starrysky.playback.Playback.Companion.STATE_PLAYING
import com.lzx.starrysky.utils.isFLAC
import com.lzx.starrysky.utils.isRTMP
import com.lzx.starrysky.utils.orDef


/**
 * isAutoManagerFocus 是否让播放器自动管理焦点
 */
class ExoPlayback(
    val context: Context,
    private val cache: ICache?,
    private val isAutoManagerFocus: Boolean
) : Playback, FocusManager.OnFocusStateChangeListener {

    companion object {
        const val TYPE_RTMP = 4
        const val TYPE_FLAC = 5
    }

    private var dataSourceFactory: DataSource.Factory? = null
    private var player: SimpleExoPlayer? = null
    private var mediaSource: MediaSource? = null
    private var trackSelector: DefaultTrackSelector? = null
    private var trackSelectorParameters: DefaultTrackSelector.Parameters? = null

    private var currSongInfo: SongInfo? = null
    private var callback: Playback.Callback? = null
    private val eventListener by lazy { ExoPlayerEventListener() }
    private var sourceTypeErrorInfo: SourceTypeErrorInfo = SourceTypeErrorInfo()
    private var focusManager = FocusManager(context)
    private var hasError = false

    init {
        focusManager.listener = this
    }

    override fun playbackState(): Int {
        return if (player == null) {
            STATE_IDLE
        } else {
            when (player?.playbackState) {
                Player.STATE_IDLE -> STATE_IDLE //error或stop
                Player.STATE_BUFFERING -> STATE_BUFFERING
                Player.STATE_READY -> {
                    if (player?.playWhenReady == true) STATE_PLAYING else STATE_PAUSED
                }
                Player.STATE_ENDED -> STATE_IDLE
                else -> STATE_IDLE
            }
        }
    }

    override fun isPlaying(): Boolean = player?.playWhenReady == true

    override fun currentStreamPosition(): Long = player?.currentPosition.orDef()

    override fun bufferedPosition(): Long = player?.bufferedPosition.orDef()

    override fun duration(): Long = if (player?.duration.orDef() > 0) player?.duration.orDef() else 0

    override var currentMediaId: String = ""

    override fun setVolume(audioVolume: Float) {
        var volume = audioVolume
        if (volume < 0) {
            volume = 0f
        }
        if (volume > 1) {
            volume = 1f
        }
        player?.volume = volume
    }

    override fun getVolume(): Float = player?.volume ?: -1f

    override fun getCurrPlayInfo(): SongInfo? = currSongInfo

    override fun getAudioSessionId(): Int = player?.audioSessionId ?: 0

    private fun getPlayWhenReady() = player?.playWhenReady ?: false

    override fun play(songInfo: SongInfo, isPlayWhenReady: Boolean) {
        val mediaId = songInfo.songId
        if (mediaId.isEmpty()) {
            return
        }
        currSongInfo = songInfo
        val mediaHasChanged = mediaId != currentMediaId
        if (mediaHasChanged) {
            currentMediaId = mediaId
        }
        StarrySky.log(
            "title = " + songInfo.songName +
                    " \n音频是否有改变 = " + mediaHasChanged +
                    " \n是否立即播放 = " + isPlayWhenReady +
                    " \nurl = " + songInfo.songUrl
        )

        //url 处理
        var source = songInfo.songUrl
        if (source.isEmpty()) {
            callback?.onPlaybackError(currSongInfo, "播放 url 为空")
            return
        }
        source = source.replace(" ".toRegex(), "%20") // Escape spaces for URL
        //代理url
        val proxyUrl = cache?.getProxyUrl(source, songInfo)
        source = if (proxyUrl.isNullOrEmpty()) source else proxyUrl
        mediaSource = createMediaSource(source)
        if (mediaSource == null) return
        if (mediaHasChanged || player == null) {
            //创建播放器实例
            createExoPlayer()

            player?.setMediaSource(mediaSource!!)
            player?.prepare()
            if (!isAutoManagerFocus) {
                focusManager.updateAudioFocus(getPlayWhenReady(), STATE_BUFFERING)
            }
        }
        //当错误发生时，如果还播放同一首歌，
        //这时候需要重新加载一下，并且吧进度 seekTo 到出错的地方
        if (sourceTypeErrorInfo.happenSourceError && !mediaHasChanged) {
            player?.setMediaSource(mediaSource!!)
            player?.prepare()
            if (!isAutoManagerFocus) {
                focusManager.updateAudioFocus(getPlayWhenReady(), STATE_BUFFERING)
            }
            if (sourceTypeErrorInfo.currPositionWhenError != 0L) {
                if (sourceTypeErrorInfo.seekToPositionWhenError != 0L) {
                    player?.seekTo(sourceTypeErrorInfo.seekToPositionWhenError)
                } else {
                    player?.seekTo(sourceTypeErrorInfo.currPositionWhenError)
                }
            }
        }
        StarrySky.log("isPlayWhenReady = $isPlayWhenReady")
        StarrySky.log("---------------------------------------")
        //如果准备好就播放
        if (isPlayWhenReady) {
            player?.playWhenReady = true
            hasError = false
            if (!isAutoManagerFocus) {
                player?.playbackState?.let { focusManager.updateAudioFocus(getPlayWhenReady(), it) }
            }
        }
    }

    private fun String.hasMediaSource(): Boolean =
        runCatching {
            Class.forName("com.google.android.exoplayer2.$this")
            return@runCatching true
        }.onFailure {
            it.printStackTrace()
        }.getOrElse { false }

    @Synchronized
    private fun createMediaSource(source: String): MediaSource {
        val uri = Uri.parse(source)
        val isRtmpSource = source.isRTMP()
        val isFlacSource = source.isFLAC()
        val type = when {
            isRtmpSource -> TYPE_RTMP
            isFlacSource -> TYPE_FLAC
            else -> Util.inferContentType(uri, null)

        }
        dataSourceFactory = buildDataSourceFactory(type)
        return when (type) {
            C.TYPE_DASH -> {
                if ("source.dash.DashMediaSource".hasMediaSource()) {
                    return DashMediaSource.Factory(dataSourceFactory!!).createMediaSource(MediaItem.fromUri(uri))
                } else {
                    throw IllegalStateException("has not DashMediaSource")
                }
            }
            C.TYPE_SS -> {
                if ("source.smoothstreaming.SsMediaSource".hasMediaSource()) {
                    return SsMediaSource.Factory(dataSourceFactory!!).createMediaSource(MediaItem.fromUri(uri))
                } else {
                    throw IllegalStateException("has not SsMediaSource")
                }
            }
            C.TYPE_HLS -> {
                if ("source.hls.HlsMediaSource".hasMediaSource()) {
                    return HlsMediaSource.Factory(dataSourceFactory!!).createMediaSource(MediaItem.fromUri(uri))
                } else {
                    throw IllegalStateException("has not HlsMediaSource")
                }
            }
            C.TYPE_RTSP -> {
                if ("source.rtsp.RtspMediaSource".hasMediaSource()) {
                    return RtspMediaSource.Factory().createMediaSource(MediaItem.fromUri(uri))
                } else {
                    throw IllegalStateException("has not RtspMediaSource")
                }
            }
            C.TYPE_OTHER -> {
                ProgressiveMediaSource.Factory(dataSourceFactory!!).createMediaSource(MediaItem.fromUri(uri))
            }
            TYPE_RTMP -> {
                if ("ext.rtmp.RtmpDataSourceFactory".hasMediaSource()) {
                    val factory = RtmpDataSourceFactory()
                    return ProgressiveMediaSource.Factory(factory).createMediaSource(MediaItem.fromUri(uri))
                } else {
                    throw IllegalStateException("has not RtmpDataSourceFactory")
                }
            }
            TYPE_FLAC -> {
                val extractorsFactory = DefaultExtractorsFactory()
                ProgressiveMediaSource.Factory(dataSourceFactory!!, extractorsFactory)
                    .createMediaSource(MediaItem.fromUri(uri))
            }
            else -> throw IllegalStateException("Unsupported type: $type")
        }
    }

    @Synchronized
    private fun createExoPlayer() {
        if (player == null) {
            @ExtensionRendererMode val extensionRendererMode = DefaultRenderersFactory.EXTENSION_RENDERER_MODE_PREFER
            val renderersFactory = DefaultRenderersFactory(context)
                .setExtensionRendererMode(extensionRendererMode)

            trackSelectorParameters = ParametersBuilder(context).build()
            trackSelector = DefaultTrackSelector(context)
            trackSelector?.parameters = trackSelectorParameters as DefaultTrackSelector.Parameters

            player = SimpleExoPlayer.Builder(context, renderersFactory)
                .setTrackSelector(trackSelector!!)
                .build()

            player?.addListener(eventListener)
            player?.setAudioAttributes(AudioAttributes.DEFAULT, isAutoManagerFocus)
            if (!isAutoManagerFocus) {
                player?.playbackState?.let { focusManager.updateAudioFocus(getPlayWhenReady(), it) }
            }
        }
    }

    @Synchronized
    private fun buildDataSourceFactory(type: Int): DataSource.Factory? {
        val userAgent = Util.getUserAgent(context, "StarrySky")
        val httpDataSourceFactory = DefaultHttpDataSourceFactory(
            userAgent,
            DefaultHttpDataSource.DEFAULT_CONNECT_TIMEOUT_MILLIS,
            DefaultHttpDataSource.DEFAULT_READ_TIMEOUT_MILLIS,
            true
        )
        return if (cache?.isOpenCache() == true && cache is ExoCache && !type.isStreamingType()) {
            val upstreamFactory = DefaultDataSourceFactory(context, httpDataSourceFactory)
            buildCacheDataSource(upstreamFactory, cache.getDownloadCache())
        } else {
            DefaultDataSourceFactory(context, httpDataSourceFactory)
        }
    }

    private fun Int.isStreamingType(): Boolean {
        return this == TYPE_RTMP || this == C.TYPE_DASH || this == C.TYPE_SS || this == C.TYPE_HLS
    }

    @Synchronized
    private fun buildCacheDataSource(upstreamFactory: DataSource.Factory?, cache: Cache?): CacheDataSource.Factory? {
        return cache?.let {
            CacheDataSource.Factory()
                .setCache(it)
                .setUpstreamDataSourceFactory(upstreamFactory)
                .setFlags(CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR)
        }
    }

    override fun stop() {
        player?.stop(true)
        player?.release()
        player?.removeListener(eventListener)
//        player?.removeAnalyticsListener(analyticsListener)
        player = null
        if (!isAutoManagerFocus) {
            focusManager.release()
        }
    }

    override fun pause() {
        player?.playWhenReady = false
        if (!isAutoManagerFocus) {
            player?.playbackState?.let { focusManager.updateAudioFocus(getPlayWhenReady(), it) }
        }
    }

    override fun seekTo(position: Long) {
        player?.seekTo(position)
        sourceTypeErrorInfo.seekToPosition = position
        if (sourceTypeErrorInfo.happenSourceError) {
            sourceTypeErrorInfo.seekToPositionWhenError = position
        }
    }

    override fun onFastForward(speed: Float) {
        player?.let {
            val currSpeed = it.playbackParameters.speed
            val currPitch = it.playbackParameters.pitch
            val newSpeed = currSpeed + speed
            it.setPlaybackParameters(PlaybackParameters(newSpeed, currPitch))
        }
    }

    override fun onRewind(speed: Float) {
        player?.let {
            val currSpeed = it.playbackParameters.speed
            val currPitch = it.playbackParameters.pitch
            var newSpeed = currSpeed - speed
            if (newSpeed <= 0) {
                newSpeed = 0f
            }
            it.setPlaybackParameters(PlaybackParameters(newSpeed, currPitch))
        }
    }

    override fun onDerailleur(refer: Boolean, multiple: Float) {
        player?.let {
            val currSpeed = it.playbackParameters.speed
            val currPitch = it.playbackParameters.pitch
            val newSpeed = if (refer) currSpeed * multiple else multiple
            if (newSpeed > 0) {
                it.setPlaybackParameters(PlaybackParameters(newSpeed, currPitch))
            }
        }
    }

    override fun getPlaybackSpeed(): Float {
        return player?.playbackParameters?.speed ?: 1.0f
    }

    override fun skipToNext() {
        callback?.skipToNext()
    }

    override fun skipToPrevious() {
        callback?.skipToPrevious()
    }

    override fun setCallback(callback: Playback.Callback?) {
        this.callback = callback
    }

    private inner class ExoPlayerEventListener : Player.EventListener {

        override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
            var newState = STATE_IDLE
            when (playbackState) {
                Player.STATE_IDLE -> {
                    //error和stop的时候会是这个状态，这里过滤掉error，避免重复回调
                    newState = if (hasError) STATE_ERROR else STATE_IDLE
                }
                Player.STATE_READY -> {
                    newState = if (player?.playWhenReady == true) STATE_PLAYING else STATE_PAUSED
                }
                Player.STATE_ENDED -> newState = STATE_IDLE
                Player.STATE_BUFFERING -> newState = STATE_BUFFERING
            }
            if (!hasError) {
                callback?.onPlayerStateChanged(currSongInfo, playWhenReady, newState)
            }
            if (playbackState == Player.STATE_READY) {
                sourceTypeErrorInfo.clear()
            }
            if (playbackState == Player.STATE_IDLE) {
                currentMediaId = ""
            }
        }

        override fun onPlayerError(error: ExoPlaybackException) {
            error.printStackTrace()
            hasError = true
            val what: String = when (error.type) {
                ExoPlaybackException.TYPE_SOURCE -> error.sourceException.message.toString()
                ExoPlaybackException.TYPE_RENDERER -> error.rendererException.message.toString()
                ExoPlaybackException.TYPE_UNEXPECTED -> error.unexpectedException.message.toString()
                else -> "Unknown: $error"
            }
            if (error.type == ExoPlaybackException.TYPE_SOURCE) {
                sourceTypeErrorInfo.happenSourceError = true
                sourceTypeErrorInfo.seekToPositionWhenError = sourceTypeErrorInfo.seekToPosition
                sourceTypeErrorInfo.currPositionWhenError = currentStreamPosition()
            }
            callback?.onPlaybackError(currSongInfo, "ExoPlayer error $what")
        }
    }

    override fun focusStateChange(info: FocusInfo) {
        if (isAutoManagerFocus) {
            return
        }
        callback?.onFocusStateChange(FocusInfo(currSongInfo, info.audioFocusState, info.playerCommand, info.volume))
    }
}

/**
 * 发生错误时保存的信息
 */
class SourceTypeErrorInfo {
    var seekToPosition = 0L
    var happenSourceError = false //是否发生资源问题的错误
    var seekToPositionWhenError = 0L
    var currPositionWhenError = 0L //发生错误时的进度

    fun clear() {
        happenSourceError = false //是否发生资源问题的错误
        seekToPosition = 0L
        seekToPositionWhenError = 0L
        currPositionWhenError = 0L //发生错误时的进度
    }
}

